integrations/tensorflow/e2e/math/math_test.py

```diff
@@ -0,0 +1,561 @@
+    'Whether or not to print out all functions with complex inputs '
+    '(and skip running the tests).')
+
+
+def _complex_wrapper(function):
+  """Wraps a tf.function to allow compiling functions of complex numbers."""
+
+  def decorator(*args, **kwargs):
+    inputs = []
+    for real, imag in zip(args[::2], args[1::2]):
+      inputs.append(tf.complex(real, imag))
+    result = function(*inputs, **kwargs)
+    # Directly returning a complex number causes an error.
+    return tf.math.real(result) + tf.math.imag(result)
+
+  return decorator
+
+
+def convert_if_complex(function, config: Config):
+  """Compatibility layer for testing complex numbers."""
+  if is_complex(config):
+    if not all([spec.dtype.is_complex for spec in config.signature]):
+      raise NotImplementedError("Signatures with mixed complex and non-complex "
+                                "tensor specs are not supported.")
+
+    # Rewrite the signature, replacing all complex tensors with pairs of real
+    # and imaginary tensors.
+    signature = []
+    for spec in config.signature:
+      new_dtype = tf.float32 if spec.dtype.size == 8 else tf.float64
+      signature.append(tf.TensorSpec(spec.shape, new_dtype))
+      signature.append(tf.TensorSpec(spec.shape, new_dtype))
+
+    return _complex_wrapper(function), signature
+  else:
+    return function, config.signature
+
+
+def _make_dims_dynamic(spec: tf.TensorSpec) -> tf.TensorSpec:
+  return tf.TensorSpec([None] * len(spec.shape), spec.dtype)
+
+
+def create_function_unittest(config: Config, function_name: str,
+                             exported_name: str) -> tf.function:
+  """Creates a tf_function_unittest from the provided Config."""
+  tf_math_function = getattr(tf.math, function_name)
+  tf_math_function, signature = convert_if_complex(tf_math_function, config)
+  wrapped_function = lambda *args: tf_math_function(*args, **config.kwargs)
+
+  if FLAGS.dynamic_dims:
+    signature = tf_utils.apply_function(signature, _make_dims_dynamic)
+
+  return tf_test_utils.tf_function_unittest(
+      input_signature=signature,
+      input_generator=config.generator,
+      input_args=config.args,
+      name=exported_name)(wrapped_function)
+
+
+def create_unittests(module: Union["TfMathModule", Type["TfMathModule"]]):
+  """Creates all tf_function_unittests to test and puts them on module."""
+  for function_name in FLAGS.functions:
+    # pytype: disable=attribute-error :\
+    for exported_name, config in FUNCTION_TO_CONFIGS[function_name].items():
+      # pytype: enable=attribute-error :/
+      if is_complex(config) and not FLAGS.test_complex:
+        continue
+      function_unittest = create_function_unittest(config, function_name,
+                                                   exported_name)
+      setattr(module, exported_name, function_unittest)
+
+
+class TfMathModule(tf_test_utils.TestModule):
+
+  @classmethod
+  def configure_class(cls):
+    create_unittests(cls)
```

Comment:
Why do we need create_unittests in configure_class and `__init__`?

Reply:
_Something_ strips the attributes that we add to the `tf.function` at module class initialization, so we can only access these attributes from the functions stored on the module class itself. I don't know what does this but it's pretty annoying :/

I would subclass the tf.function, but tensorflow doesn't want you to do that (e.g. `tf.python` isn't available to the user in all installations of tensorflow. I also tried and it does weird things when you try to subclass it). The options are:

1. store the unittest information as global state in `tf_test_utils.py`, similar to how we do that for the compiled modules.
2. require that the `tf_function_unittest`s exist on the cls

Reply:
I decided to implement it with global state so that the rest of the code base would be less cumbersome / confusing. The implementation itself wasn't bad, and only `tf_test_utils` itself should ever have to interact with `_global_unittest_configs`.